File: Examples/Objective-C/FileShare/FileShare/FilesTableViewController.m
//
//  FilesTableViewController.m
//  Shared Files
//
//  Created by Calvin on 6/7/17.
//  Copyright © 2017 Bridgefy Inc. All rights reserved.
//

#import "FilesTableViewController.h"

NSString *const localUser = @"local user";

@interface FilesTableViewController () <UIDocumentPickerDelegate, UIImagePickerControllerDelegate, UINavigationControllerDelegate>

@end

@implementation FilesTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    if (self.peer) {
        self.title = [self.peer formattedName];
        
    } else {
        self.title = @"Shared files";
        
        // When looking for local files an "add file" button is required
        UIBarButtonItem *addFileButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd
                                                                                       target:self
                                                                                       action:@selector(showAddLocalFileOptions)];
        self.navigationItem.rightBarButtonItem = addFileButton;
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (self.peer) {
        if (self.peer.files.count == 0) {
            [self showEmptyMessage];
        } else {
            [self removeEmptyMessage];
        }
        return self.peer.files.count;
    } else {
        if (self.files.count == 0) {
            [self showEmptyMessage];
        } else {
            [self removeEmptyMessage];
        }
        return self.files.count;
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"fileCell"
                                                            forIndexPath:indexPath];
    
    UILabel *fileNameLabel = [(UILabel *)cell.contentView viewWithTag:1000];
    UILabel *fileSizeLabel = [(UILabel *)cell.contentView viewWithTag:1001];
    UIButton *downloadButton = [(UIButton *)cell.contentView viewWithTag:1002];
    UIActivityIndicatorView *activityIndicator = (UIActivityIndicatorView *)[cell.contentView viewWithTag:1003];
    
    FileInfo *fileInfo;
    
    if (self.peer) {
        fileInfo = [self.peer.files objectAtIndex:indexPath.row];
    } else {
        fileInfo = [self.files objectAtIndex:indexPath.row];
    }
    
    fileNameLabel.text = fileInfo.name;
    fileSizeLabel.text = [fileInfo formattedFileSize];
    
    downloadButton.hidden = fileInfo.local;
    [activityIndicator stopAnimating];
    
    return cell;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    return self.peer == nil;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        if (self.delegate) {
            if ([self.delegate deleteFileInfo:(FileInfo *)[self.files objectAtIndex:indexPath.row]]) {
                [self.tableView deleteRowsAtIndexPaths:@[indexPath]
                                      withRowAnimation:UITableViewRowAnimationAutomatic];
            }
            
        }
    }
}

- (void)showEmptyMessage {
    UILabel *emptyMessageLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, self.view.bounds.size.width, self.view.bounds.size.height)];
    emptyMessageLabel.text = @"No shared files";
    emptyMessageLabel.textColor = [UIColor blackColor];
    emptyMessageLabel.numberOfLines = 0;
    emptyMessageLabel.textAlignment = NSTextAlignmentCenter;
    emptyMessageLabel.font = [UIFont systemFontOfSize:17];
    [emptyMessageLabel sizeToFit];
    
    self.tableView.backgroundView = emptyMessageLabel;
    self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
}

- (void)removeEmptyMessage {
    self.tableView.backgroundView = nil;
    self.tableView.separatorStyle = UITableViewCellSeparatorStyleSingleLine;
}

- (void)showAddLocalFileOptions {
    UIAlertController *optionsController = [UIAlertController alertControllerWithTitle:@"Add new file to share"
                                                                               message:@"Select the source"
                                                                        preferredStyle:UIAlertControllerStyleActionSheet];
    
    UIAlertAction *iCloudAction = [UIAlertAction actionWithTitle:@"iCloud"
                                                           style:UIAlertActionStyleDefault
                                                         handler:^(UIAlertAction * _Nonnull action) {
                                                             [self addFileFromCloud];
                                                         }];
    
    UIAlertAction *imagesAction = [UIAlertAction actionWithTitle:@"Images"
                                                           style:UIAlertActionStyleDefault
                                                         handler:^(UIAlertAction * _Nonnull action) {
                                                             [self addFileFromImages];
                                                         }];
    
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel"
                                                           style:UIAlertActionStyleCancel
                                                         handler:nil];
    
    // To avoid
    [optionsController addAction:iCloudAction];
    [optionsController addAction:imagesAction];
    [optionsController addAction:cancelAction];
    
    [self presentViewController:optionsController
                       animated:YES
                     completion:nil];
}

- (void)addFileFromCloud {
    NSArray *documentTypes = @[@"public.content"];
    UIDocumentPickerViewController *documentPicker = [[UIDocumentPickerViewController alloc] initWithDocumentTypes:documentTypes
                                                                                                            inMode:UIDocumentPickerModeImport];
    documentPicker.modalPresentationStyle = UIModalPresentationPopover;
    documentPicker.delegate = self;
    
    [self presentViewController:documentPicker
                       animated:YES
                     completion:nil];
    
}

- (void)addFileFromImages {
    UIImagePickerController *imagePicker = [[UIImagePickerController alloc] init];
    
    imagePicker.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
    imagePicker.mediaTypes = @[(NSString *)kUTTypeImage];
    imagePicker.delegate = self;
    
    [self presentViewController:imagePicker
                       animated:YES
                     completion:nil];
}

#pragma mark - Image picker delegate

- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info {
    
    [picker dismissViewControllerAnimated:YES
                               completion:nil];
    
    if (self.delegate) {
        [self.delegate createNewFileFromMediaInfo:info];
    }
}

#pragma mark - Document Picker delegate

- (void)documentPicker:(UIDocumentPickerViewController *)controller didPickDocumentAtURL:(NSURL *)url {
    if (self.delegate) {
        [self.delegate createNewFileFromCloudPath:url.path];
    }
}

- (void)updateAvailableFiles {
    [self.tableView reloadData];
}

- (IBAction)downloadButtonPressed:(id)sender {
    CGPoint buttonPosition = [sender convertPoint:CGPointZero toView:self.tableView];
    NSIndexPath *ip = [self.tableView indexPathForRowAtPoint:buttonPosition];
    
    UIButton *downloadButton = (UIButton *)sender;
    downloadButton.hidden = YES;
    
    UITableViewCell *cell = [self.tableView cellForRowAtIndexPath:ip];
    UIActivityIndicatorView *activityIndicator = (UIActivityIndicatorView *)[cell.contentView viewWithTag:1003];
    [activityIndicator startAnimating];
    activityIndicator.hidden = NO;
    
    FileInfo *file = [self.peer.files objectAtIndex:ip.row];
    
    if (self.delegate) {
        [self.delegate requestFile:file.uuid fromPeer:self.peer];
    }
}

@end

File: Examples/Swift/FileShare/FileShare/FilesTableViewController.swift
//
//  FilesTableViewController.swift
//  FileShare
//
//  Created by Calvin on 6/16/17.
//  Copyright © 2017 Bridgefy Inc. All rights reserved.
//

import UIKit
import MobileCoreServices

protocol FilesTableViewControllerDelegate: class {
    func createNewFile(mediaInfo info: [String: Any])
    func createNewFile(cloudPath path: String)
    func requestFile(_ id: String, _ peer: Peer)
    func deleteFile(fileInfo: FileInfo) -> Bool
    
}

class FilesTableViewController: UITableViewController, UIDocumentPickerDelegate, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    
    var peer: Peer?
    var files: [FileInfo] = []
    weak var delegate: FilesTableViewControllerDelegate?

    override func viewDidLoad() {
        super.viewDidLoad()
        
        guard self.peer != nil else {
            self.title = "Shared files"
            let addFileButton = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(FilesTableViewController.showAddLocalFileOptions))
            self.navigationItem.setRightBarButton(addFileButton, animated: false)
            return
        }
        
        self.title = self.peer!.formattedName()
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        guard self.peer != nil else {
            if self.files.count == 0 {
                self.showEmptyMessage()
            } else {
                self.removeEmptyMessage()
            }
            
            return self.files.count
        }
        
        if self.peer!.files.count == 0 {
            self.showEmptyMessage()
        } else {
            self.removeEmptyMessage()
        }
        
        return self.peer!.files.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "fileCell", for: indexPath)

        let fileNameLabel = cell.contentView.viewWithTag(1000) as! UILabel
        let fileSizeLabel = cell.contentView.viewWithTag(1001) as! UILabel
        let downloadButton = cell.contentView.viewWithTag(1002) as! UIButton
        let activityIndicator = cell.contentView.viewWithTag(1003) as! UIActivityIndicatorView

        let fileInfo: FileInfo
        
        if self.peer != nil {
            fileInfo = self.peer!.files[indexPath.row]
        } else {
            fileInfo = self.files[indexPath.row]
        }
        
        fileNameLabel.text = fileInfo.name
        fileSizeLabel.text = fileInfo.formattedFileSize()
        
        downloadButton.isHidden = fileInfo.local
        activityIndicator.stopAnimating()
        
        return cell
    }

    // Override to support conditional editing of the table view.
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return self.peer == nil
    }

    // Override to support editing the table view.
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            if self.delegate != nil {
                if (self.delegate?.deleteFile(fileInfo: self.files[indexPath.row]))! {
                    tableView.deleteRows(at: [indexPath], with: .fade)
                }
            }
        }
    }
    
    func showEmptyMessage() {
        let emptyMessageLabel = UILabel(frame: CGRect(x: 0, y: 0, width: self.view.bounds.size.width, height: self.view.bounds.size.height))
        emptyMessageLabel.text = "No shared files"
        emptyMessageLabel.textColor = #colorLiteral(red: 0, green: 0, blue: 0, alpha: 1)
        emptyMessageLabel.numberOfLines = 0
        emptyMessageLabel.textAlignment = .center
        emptyMessageLabel.font = UIFont.systemFont(ofSize: 17)
        emptyMessageLabel.sizeToFit()
        
        self.tableView.backgroundView = emptyMessageLabel
        self.tableView.separatorStyle = .none
    }
    
    func removeEmptyMessage() {
        self.tableView.backgroundView = nil
        self.tableView.separatorStyle = .singleLine
    }
    
    func showAddLocalFileOptions() {
        let optionsController = UIAlertController(title: "Add new file to share", message: "Select the source", preferredStyle: .actionSheet);
        
        let iCloudAction = UIAlertAction(title: "iCloud", style: .default) { (action) in
            self.addFileFromCloud()
        }
        
        let imagesAction = UIAlertAction(title: "Images", style: .default) { (action) in
            self.addFileFromImages()
        }
        
        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
        
        // To avoid
        optionsController.addAction(iCloudAction)
        optionsController.addAction(imagesAction)
        optionsController.addAction(cancelAction)
        
        self.present(optionsController, animated: true)
        
    }
    
    func addFileFromCloud() {
        let documentPicker = UIDocumentPickerViewController(documentTypes: ["public.content"], in: .import)
        documentPicker.modalPresentationStyle = .popover
        documentPicker.delegate = self;
        
        self.present(documentPicker, animated: true, completion: nil)
    }
    
    func addFileFromImages() {
        let imagePicker = UIImagePickerController()
        imagePicker.sourceType = .photoLibrary
        imagePicker.mediaTypes = [kUTTypeImage as String]
        imagePicker.delegate = self
        
        self.present(imagePicker, animated: true, completion: nil)
        
    }
    
    // MARK: - Document Picker Delegate
    
    func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentAt url: URL) {
        self.delegate?.createNewFile(cloudPath: url.path)
    }
    
    // MARK: - Image Picker Delegate
    
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
        picker.dismiss(animated: true, completion: nil)
        
        self.delegate?.createNewFile(mediaInfo: info)
    }
    
    func updateAvailableFiles() {
        self.tableView.reloadData()
    }
    
    @IBAction func dowloadButtonPressed(_ sender: Any) {
        let buttonPosition = (sender as! UIButton).convert(CGPoint.zero, to: self.tableView)
        let indexPath = self.tableView.indexPathForRow(at: buttonPosition)
        
        let downloadButton = sender as! UIButton
        downloadButton.isHidden = true
        
        let cell = self.tableView.cellForRow(at: indexPath!)
        let activityIndicator = cell?.contentView.viewWithTag(1003) as! UIActivityIndicatorView
        activityIndicator.startAnimating()
        activityIndicator.isHidden = false
        
        let fileInfo = self.peer?.files[(indexPath?.row)!]
        
        self.delegate?.requestFile((fileInfo?.uuid)!, self.peer!)
    }
    

}
